### PR TITLE
compiling.rst: multiple improvements and clarifications

### PR DESCRIPTION
**Pull request by anatoli26**
The changes are as follows:
 1. Completed the descriptions of a number of libraries with the explanations of when they are needed and what happens if they are not included (based on the information from this thread). For this I also added to the tables a new column "Required".
 2. Completed the libs package names on Debian & RH for all libs (except for the perl packages).
 3. Reordered some libs, in particular brotli, nghttp2, shapelib, wslay, etc. (they were in Other, moved them to DAV/JMAP/httpd and renamed a bit that section).
 4. Added links to the libs official websites where they were missing.
 5. Other minor cleanup.
 6. Removed some libs that IMO are not used anymore:
    1. net-snmp: in my understanding, it was used for the SNMP code in master which is pending removal (#1765); so no need to include obsolete features in the description;
    2. tcp_wrappers: not 100% sure, but it looks like it's not used anymore.

### CODE REVIEW

**Comment by elliefm** [comment]
Thanks, that's looking good!

One thing though, some of the modules are listed as "Required: yes" and "Required for make check: no" (e.g. libxml), which is nonsensical -- if you can't build without them, then you also can't run the test suite without them.

Actually, I'm not sure why we even have a "Required for make check" column?  It doesn't really matter for the majority of this stuff -- so it's just another piece of data that needs to be kept up-to-date (and probably isn't).

Some of the tables only have this column, I think those can probably just be converted to "Required" with the same value for now.  It might need a proper review at some point.

Usually you would configure cyrus with the dependencies and features you need, and then run `make check` with that configuration.  It doesn't make any sense to add dependencies _just_ to make the unit tests build/pass.

I think the only exception to this is cunit, which without this dependency we obviously we can't build our cunit tests at all.  But instead of having an extra column through the entire document, we could just move that information to the notes column... something like "Required: no.  Notes:  You need this if you wish to run `make check`"

If it turns out there are cunit tests that fail to build or run when some feature dependency is missing, then I think that's a bug and those unit tests should be made conditional -- it doesn't make sense for a disabled feature to cause test failures.  We just shouldn't run tests for features that haven't been compiled in.

**Comment by anatoli26** [comment]
Ellie,

Thanks for your feedback.

I fully agree with you with respect to the `make test` column, I just wasn't sure why it was included in the first place.

Now that you confirmed it's not actually needed, I made changes to remove it, add "Required" where it wasn't and to add a clarification for CUnit.

Please check the "Developers only" section. I'm not sure if the following dependencies are required and if the Notes for them are up-to-date: libdb-dev, perl-devel, the perl & python libs, xxd.

Please let me know what you think about the new version.

File: docsrc/imap/developer/compiling.rst [hunk 1]
```diff
@@ -275,7 +276,7 @@ via configure.
 
 Sieve is enabled by default.
 
-CalDAV, CardDAV, JMAP
+CalDAV, CardDAV, WebDAV, JMAP
```
Comment (elliefm) [inline docsrc/imap/developer/compiling.rst:279]: Heads up, sphinx is really picky about its underlines!

```
/home/ellie/fastmail/cyrus-imapd/docsrc/imap/developer/compiling.rst:328: WARNING: Title underline too short.

CalDAV, CardDAV, WebDAV, JMAP
#####################
```

**Comment by elliefm** [comment]
Thanks, that's looking good!

* libdb-dev:  not used anymore. I think 2.5 was the last version to need libdb?  Just remove this row entirely :)
* perl-devel: yep, this is needed to build the binary perl modules.  I would remove the reference to `--with-perl` as that option is junk (#2887 )
* perl and python modules: these look up to date to me, but ditto re `--with-perl`
* xxd: yep, still needed

I notice that the Notes column is quite long for many of the "Developers" dependencies, and a lot of those long notes kinda re-explain the same thing in different words.  They could be made shorter if we just added a short paragraph in between the "Developers" heading and the start of the table, perhaps something like:

> The developer dependencies are required if you are building from git sources, have modified certain source files from the release tarball, or have configured with `--enable-maintainer-mode` in order to build a new package.  
> 
> If you are building normally from a pure release tarball, then you don't need these dependencies.  The files these dependencies produce have been pre-built and included in the release, and do not normally need to be rebuilt.

Though, that being the case, I would then move the CUnit row down into the "Other" table instead -- since it's not required even when building from git (unlike the other Developers deps), but "make check" is useful even when you're building from a release tarball, and we don't/can't simply precompile the files for inclusion in the release (like we can with xxd or the documentation, etc).

**Comment by anatoli26** [comment]
Ellie,

I've incorporated all your suggested changes.

Except for transfig, all Notes in Developers section are one-liners now. And transfig should be removed one day, so its row would be gone too.

File: docsrc/imap/developer/compiling.rst [hunk 2]
```diff
@@ -112,46 +121,65 @@ Alternate database formats
...
-    `wslay`_,,,, "WebSockets support in httpd"
-    `nghttp2`_, libnghttp2-dev, libnghttp2-devel, "no", "HTTP/2 support for httpd"
+    `pcre`_, libpcre3-dev, pcre-devel, "yes", "no", "PCRE 1 (8.x) - for utf-8/unicode regular expression
+    matching. Also see `issues/1731`_."
+    `perl(Term::ReadLine)`_,,, "no", "no", "Perl library needed by **cyradm**"
+    `libsrs2`_, *no package*, *no package*, "no", "no", "It is used for
+    implementing Sender Rewriting Scheme (SRS) functionality for messages
+    forwarded by sieve scripts. Without it, messages forwarded by sieve scripts
+    will not have this functionality and might have difficulty delivering to
+    SMTP servers that insist on it."
```
[the remark this answers is not in the document]
Comment (anatoli26) [reply docsrc/imap/developer/compiling.rst:194]: 🤷‍♂️ Don't know. I would remove this feature entirely, it looks like it's not used much, complex to setup and it's unclear what servers may require it. The dependency itself is not in the major distros, should be built manually.

[the remark this answers is not in the document]
Comment (elliefm) [reply docsrc/imap/developer/compiling.rst:194]: Presumably the admin installing Cyrus knows what their outgoing SMTP server requires well enough to decide for themselves whether they need this feature or not.  So I think it's fine for us to just say that some SMTP servers may insist on it, without needing to list which ones do under which circumstances.

**Comment by elliefm** [comment]
Thanks!

I've now merged this manually, so that I could rebase out the "Sync with the master" merge commit first.  (Though, having done it manually, I've now noticed the GitHub UI provides a "Rebase and merge" option that would've done the same thing for me.  Good to know!)

As a general tip, if you need to bring your feature branch up to date with master, it's better to rebase it than to merge it.  When you merge master into a feature branch, and later merge the feature branch back into master, the history graph gets messy and hard to follow.  Rebasing instead keeps the graph shallow, which is kind to future maintainers who might need to go spelunking.

You don't usually need to bring it up to date at all, though -- only if it would conflict, which the PR will tell you about.

File: docsrc/imap/developer/compiling.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (dilyanpalauzov) [reply docsrc/imap/developer/compiling.rst:194]: I do not think that a receiving SMTP server can require SRS.  As there is no fixed schema how to do the SRS, a receiving server cannot know, if SRS in involved.  The rewriting is solely in the interest of the sending SMTP server (so that the sending server gets any bounces).
